File: backend/routes/fhir.py
"""FHIR routes – export validated data as FHIR resources."""
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from database import get_db, FHIRExport
from tier3.fhir_generator import generate_fhir_bundle

router = APIRouter()


@router.get("/fhir/{patient_id}")
async def get_fhir_bundle(patient_id: int, db: Session = Depends(get_db)):
    """Generate and return FHIR Bundle for a patient."""
    bundle = generate_fhir_bundle(patient_id, db)
    if not bundle.get("entry"):
        raise HTTPException(status_code=404, detail="Patient not found or no valid observations")
    return bundle


@router.get("/fhir/{patient_id}/download")
async def download_fhir(patient_id: int, db: Session = Depends(get_db)):
    """Download FHIR Bundle as JSON file."""
    bundle = generate_fhir_bundle(patient_id, db)
    if not bundle.get("entry"):
        raise HTTPException(status_code=404, detail="Patient not found")

    return JSONResponse(
        content=bundle,
        headers={
            "Content-Disposition": f'attachment; filename="careai_fhir_patient_{patient_id}.json"'
        },
    )


@router.get("/fhir/exports/list")
async def list_exports(db: Session = Depends(get_db)):
    """List all FHIR exports."""
    exports = db.query(FHIRExport).order_by(FHIRExport.created_at.desc()).all()
    return {
        "exports": [
            {
                "id": e.id,
                "patient_id": e.patient_id,
                "resource_type": e.resource_type,
                "version": e.version,
                "created_at": e.created_at.isoformat() if e.created_at else None,
            }
            for e in exports
        ]
    }

File: backend/routes/fhir.js
/**
 * FHIR Export API Routes
 */
const express = require('express');
const router = express.Router();
const { parsePositiveInt, route, sendJson } = require('../lib/http');
const { generateBundle, getDownloadBundle } = require('../services/fhir-service');

// GET /api/fhir/:patientId — generate FHIR bundle for a patient
router.get('/:patientId', route((req, res) => {
  const patientId = parsePositiveInt(req.params.patientId, { field: 'patient id' });
  sendJson(res, generateBundle(patientId, { minDataQuality: req.query.min_dq }));
}));

// GET /api/fhir/:patientId/download — download FHIR bundle as file
router.get('/:patientId/download', route((req, res) => {
  const patientId = parsePositiveInt(req.params.patientId, { field: 'patient id' });
  const { patient, bundle } = getDownloadBundle(patientId);
  res.setHeader('Content-Disposition', `attachment; filename="fhir-bundle-${patient.medical_record_number}.json"`);
  res.setHeader('Content-Type', 'application/fhir+json');
  res.json(bundle);
}));

module.exports = router;
